### PR TITLE
Add CUDA demos

### PR DESCRIPTION
**Pull request by kchristin22**
This PR adds the following demos:
* custom TensorContraction demo
* NVIDIA's Black-Scholes sample

### CODE REVIEW

File: demos/CUDA/BlackScholes/BlackScholes.cu [hunk 1]
```diff
@@ -0,0 +1,405 @@
...
+
+/*
+ * This sample evaluates fair call and put prices for a
+ * given set of European options by Black-Scholes formula.
+ * See supplied whitepaper for more explanations.
+ */
+
+/*
+ * DISCLAIMER: The following file has been modified slightly to make it
+ * compatible with Clad. The original file can be found at NVIDIA's cuda-samples
```
Comment (vgvassilev) [inline demos/CUDA/BlackScholes/BlackScholes.cu:36]: Can we describe what the modifications are on a high-level?

File: demos/CUDA/TensorContraction.cu [hunk 2]
```diff
@@ -0,0 +1,266 @@
+//--------------------------------------------------------------------*- C++ -*-
```
Comment (vgvassilev) [inline demos/CUDA/TensorContraction.cu:1]: Can we get this demo in a separate PR?

File: demos/CUDA/BlackScholes/BlackScholes.cu (hunk 1 above)
Comment (kchristin22) [reply demos/CUDA/BlackScholes/BlackScholes.cu]: Hmm I can mention that these changes refer to the original `main` function whose parts have been moved to a separate function to use `clad::gradient` on. There's also the loss of the original print statements and the addition of helper functions to verify the results. Would this do as a whole?

File: demos/CUDA/TensorContraction.cu (hunk 2 above)
Comment (kchristin22) [reply demos/CUDA/TensorContraction.cu:1]: Certainly

File: demos/CUDA/BlackScholes/BlackScholes.cu (hunk 1 above)
Comment (kchristin22) [reply demos/CUDA/BlackScholes/BlackScholes.cu]: How about now?

Comment (vgvassilev) [reply demos/CUDA/BlackScholes/BlackScholes.cu]: Yes, or just make a diff and put it in comment if it’s not big.

Comment (kchristin22) [reply demos/CUDA/BlackScholes/BlackScholes.cu]: I think that would be confusing, because I kind of restructured the file

**Comment by kchristin22** [comment]
I added as many parts as possible from the original demo to minimize the diff. You're free to have a look in case you have any suggestions!

**Comment by kchristin22** [comment]
@vgvassilev Let's postpone this merge till the r value PR is fixed and merged, as this functionality is needed to run the demo

**Review by vgvassilev — APPROVED** [review]
LGTM!

**Pull request merged**